File: grpc/go/src/cheetah/client_init.go
/*
 * Copyright (c) 2016 by cisco Systems, Inc.
 * All rights reserved.
 */
package access_point

/* Standard packages */
import (
    "fmt"
    "log"
    "google.golang.org/grpc"
    "golang.org/x/net/context"
)

/* Cheetah packages */
import (
    pb "gengo"
)

func ClientInit(conn *grpc.ClientConn) (int) {
    /* Setup a go-routine channel to synchronize both go-routines*/
    sync_chan := make(chan int)

    /* Setup the notification channel */
    go setupNotifChannel(conn, sync_chan)

    /* Wait for response 0: error. 1: all ok*/
    wait_resp := <- sync_chan
    if (wait_resp == 0) {
        fmt.Println("Client Error")
        return 0
    }

    /* Create a sLGlobalClient instance */
    globalClient := pb.NewAPGlobalClient(conn)

    /* Create a APGlobalsGetMsg */
    globalGetMsg := &pb.APGlobalsGetMsg {}

    /* RPC to Get the globals. */
    response, err := globalClient.APGlobalsGet(context.Background(),
        globalGetMsg)
    if (err != nil) {
        fmt.Println("Client Error %v", err)
        return 0
    }

    /* Print Server response */
    fmt.Println("Server response: ", response.ErrStatus.Status)
    fmt.Println("Max Radio Name Len  : ", response.MaxRadioNameLength)
    fmt.Println("Max Ssid Name Len   : ", response.MaxSsidNameLength)

    return wait_resp
}

func setupNotifChannel(conn *grpc.ClientConn, sync_chan chan int) {
    /* Create a sLGlobalClient instance */
    globalClient := pb.NewAPGlobalClient(conn)

    /* Create a APGlobalsGetMsg */
    initMsg := &pb.APInitMsg {
        MajorVer: uint32(pb.APVersion_AP_MAJOR_VERSION),
        MinorVer: uint32(pb.APVersion_AP_MINOR_VERSION),
        SubVer: uint32(pb.APVersion_AP_SUB_VERSION),
    }

    /* RPC to Init the notification channel */
    stream, err := globalClient.APGlobalInitNotif(context.Background(),
        initMsg)
    if err != nil {
        fmt.Println("Client Error %v", err)
        /*signal error*/
        sync_chan <- 0
        return
    }

    /* For ever read from stream */
    for {
        event, stream_err := stream.Recv()
        if (stream_err != nil) {
            fmt.Println("Client Recv Error %v", stream_err)
            break
        }

        switch event.EventType {
        case pb.APGlobalNotifType_AP_GLOBAL_EVENT_TYPE_VERSION:
            //initMsgRsp := event.GetInitRspMsg()
            /* Check Server event */
            if (event.ErrStatus.Status == pb.APErrorStatus_AP_SUCCESS) ||
               (event.ErrStatus.Status ==
                   pb.APErrorStatus_AP_INIT_STATE_CLEAR) ||
               (event.ErrStatus.Status ==
                   pb.APErrorStatus_AP_INIT_STATE_READY) {
                fmt.Printf("Server Returned %s\n",
                    event.ErrStatus.Status.String());
                /*fmt.Printf("Server Returned 0x%x, Version: %d.%d.%d\n",
                    event.ErrStatus.Status,
                    initMsgRsp.MajorVer, initMsgRsp.MinorVer, initMsgRsp.SubVer) */
                /*signal success, continue processing events from server*/
                sync_chan <- 1
            } else {
                log.Fatalf("Client Recv Error 0x%x", event.ErrStatus.Status)
            }

        case pb.APGlobalNotifType_AP_GLOBAL_EVENT_TYPE_ERROR:
            if (event.ErrStatus.Status == pb.APErrorStatus_AP_NOTIF_TERM) {
                log.Fatalf("Received notice to terminate. Client Takeover?\n")
            }
            log.Fatalf("Error not handled: 0x%x", event.ErrStatus.Status)

        case pb.APGlobalNotifType_AP_GLOBAL_EVENT_TYPE_HEARTBEAT:
            fmt.Printf("Received HeartBeat\n")

        default:
            log.Fatalf("Client Recv unknown event %s",
                event.EventType.String())
        }
    }

    log.Fatalf("Exiting")
}

File: grpc/python/src/tutorial/client_init.py
#
# Copyright (c) 2017 by Cisco Systems, Inc.
# All rights reserved.
#

# Standard python libs
import os
import sys
import threading

# Add the generated python bindings to the path
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.realpath(__file__))))

# gRPC generated python bindings
from genpy import ap_global_pb2
from genpy import ap_common_types_pb2
from genpy import ap_version_pb2

# gRPC libs
from grpc.beta import implementations

from util import util

#
# Client Init: Initialize client session
#    stub: GRPC stub
#
def client_init(stub, event):
    #
    # Create APInitMsg to handshake the version number with the server.
    # The Server will allow/deny access based on the version number.
    # The same RPC is used to setup a notification channel for global
    # events coming from the server.

    # Set the client version number based on the current proto files' version
    init_msg = ap_global_pb2.APInitMsg()
    init_msg.MajorVer = ap_version_pb2.AP_MAJOR_VERSION
    init_msg.MinorVer = ap_version_pb2.AP_MINOR_VERSION
    init_msg.SubVer = ap_version_pb2.AP_SUB_VERSION

    # Set a very large timeout, as we will "for ever" loop listening on
    # notifications from the server
    Timeout = 365*24*60*60 # Seconds

    # This for loop will never end unless the server closes the session
    for response in stub.APGlobalInitNotif(init_msg, Timeout):
        if response.EventType == ap_global_pb2.AP_GLOBAL_EVENT_TYPE_VERSION:
            if (ap_common_types_pb2.APErrorStatus.AP_SUCCESS ==
                    response.ErrStatus.Status) or \
                (ap_common_types_pb2.APErrorStatus.AP_INIT_STATE_CLEAR ==
                    response.ErrStatus.Status) or \
                (ap_common_types_pb2.APErrorStatus.AP_INIT_STATE_READY ==
                    response.ErrStatus.Status):
                print "Server Returned 0x%x, Version %d.%d.%d" %(
                    response.ErrStatus.Status,
                    response.InitRspMsg.MajorVer,
                    response.InitRspMsg.MinorVer,
                    response.InitRspMsg.SubVer)
                print "Successfully Initialized, connection established!"
                # Any thread waiting on this event can proceed
                event.set()
            else:
                print "client init error code 0x%x", response.ErrStatus.Status
                os._exit(0)
        elif response.EventType == ap_global_pb2.AP_GLOBAL_EVENT_TYPE_HEARTBEAT:
            print "Received HeartBeat"
        elif response.EventType == ap_global_pb2.AP_GLOBAL_EVENT_TYPE_CONFIG:
            print "Received Config"
            print "   Token: %s" %(response.CfgRspMsg.Token)
            print "   ProxyURL: %s" %(response.CfgRspMsg.ProxyURL)
            print "   ProxyPort: %d" %(response.CfgRspMsg.ProxyPort)
        elif response.EventType == ap_global_pb2.AP_GLOBAL_EVENT_TYPE_ERROR:
            if (ap_common_types_pb2.APErrorStatus.AP_NOTIF_TERM ==
                    response.ErrStatus.Status):
                print "Received notice to terminate. Client Takeover?"
                os._exit(0)
            else:
                print "Error not handled:", response
        else:
            print "client init unrecognized response %d", response.EventType
            os._exit(0)

#
# Thread starting point
#
def global_thread(stub, event):
    print "Global thread spawned"

    # Initialize the GRPC session. This function should never return
    client_init(stub, event)

    print "global_thread: exiting unexpectedly"
    # If this session is lost, then most likely the server restarted
    # Typically this is handled by reconnecting to the server. For now, exit()
    os._exit(0)

#
# Spawn a thread for global events
#
def global_init(channel):
    # Create the gRPC stub.
    stub = ap_global_pb2.beta_create_APGlobal_stub(channel)

    # Create a thread sync event. This will be used to order thread execution
    event = threading.Event()

    # The main reason we spawn a thread here, is that we dedicate a GRPC
    # channel to listen on Global asynchronous events/notifications.
    # This thread will be handling these event notifications.
    t = threading.Thread(target = global_thread, args=(stub, event))
    t.start()

    # Wait for the spawned thread before proceeding
    event.wait()

    # Get the globals. Create a APGlobalsGetMsg
    global_get = ap_global_pb2.APGlobalsGetMsg()

    #
    # Make an RPC call to get global attributes
    #
    Timeout = 10 # Seconds
    response = stub.APGlobalsGet(global_get, Timeout)

    # Check the received result from the Server
    if (response.ErrStatus.Status ==
        ap_common_types_pb2.APErrorStatus.AP_SUCCESS):
        util.print_globals(response)
    else:
        print "Globals response Error 0x%x" %(response.ErrStatus.Status)
        os._exit(0)

#
# Setup the GRPC channel with the server, and issue RPCs
#
if __name__ == '__main__':
    from util import util
    server_ip, server_port = util.get_server_ip_port()

    print "Using GRPC Server: IP(%s) Port(%s)" %(server_ip, server_port)

    # Create the channel for gRPC.
    channel = implementations.insecure_channel(server_ip, server_port)

    # Initialize client (check major/minor versions and globals)
    global_init(channel)

    # Exit and Kill any running GRPC threads.
    os._exit(0)
